src/main/java/com/google/sps/servlets/NotificationServlet.java

```diff
@@ -0,0 +1,208 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+ 
+import com.google.sps.classes.Notification;
+import com.google.sps.classes.Utility;
+import java.io.IOException;
+import java.sql.*;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+ 
+/**
+* Servlet that handles the fetching and posting of notifications.
+*/
+@WebServlet("/notification")
+public class NotificationServlet extends HttpServlet {
+  /**
+   * Fetches the notifications of a specific user.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get user's ID from Utility method.
+    int userId = Utility.getUserId();
+    if (userId != -1) {
+      // Fetch notifications if user is signed in and convert the ArrayList to JSON
+      // using Utility method.
+      response.setContentType("application/json;");
+      response.getWriter().println(Utility.convertToJsonUsingGson(getNotifications(userId));); 
+    }
+  }
+
+  /**
+   * Posts a new notification for all followers of a modified question or answer.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
This is a very very long function. Think through how to split it up into some helper functions! Note that helper functions can either be private methods in this file, or can go in Utility if they will be or could be utilized in multiple files.

Reply:
You are right! I broke up logic inside of if...else... block. Also separated the insertion into helper function. Can you see anyway of further simplifying this code.

Reply:
This is much better, thank you! Try to keep your functions at a max ~40 lines, though complicated functions can of course get longer, so it's not a hard rule.